### ISSUE TITLE
[rfc] Elastic training using AdaptDL

### ISSUE THREAD

**Issue by odp** [ray-project/ray#17054]
# Elastic training using AdaptDL

| Status        | Proposed                                             |
:-------------- |:---------------------------------------------------- |
| **RFC #**     | #17054 |
| **Author(s)** | Omkar Pangarkar (@odp), Aurick Qiao (@aurickq) |
| **Sponsor**   | Richard Liaw (@richardliaw), Hao Zhang (@zhisbug) |
| **Updated**   | 2021-07-13                                           |

## Objective

AdaptDL is a DL job scheduler that automatically scales/rescales distributed (data-parallel) training jobs based on their performance characteristics (mainly computation time/step time and communication time/sync time between replicas), available resources and other jobs. Its goal is to allocate cluster resources fairly and efficiently between multiple competing jobs. AdaptDL uses the co-adaptive scheduling proposed in [this paper](https://arxiv.org/pdf/2008.12260.pdf), which means that automatic batch size scaling based on training goodput is used to increase resource efficiency. AdaptDL can also expand the cluster if the jobs could run faster with more resources as determined by the performance models that the scheduler maintains internally for each job. 

This is a proposal to integrate the above functionality with Ray in the form of a Tune trial scheduler. 

## Motivation

Currently, [the scheduler](https://github.com/petuum/adaptdl) is implemented as a custom Kubernetes scheduler which manages job CRDs. Elasticity and use of spot instances are supported on AWS. The user is required to use a few modified PyTorch APIs to use these features and then scaling is handled automatically for them. 

## User Benefit

AdaptDL will analyze multiple trials of a single experiment or multiple experiments and run them concurrently in the most resource-efficient manner (scaling them if needed) on a Ray cluster. It will also make use of the cluster expansion feature to request more resources on-the-fly. The goal is to reduce the time it takes to finish an experiment or set of experiments in the most resource and cost-effective manner. Moreover, AdaptDL can also be used as a simple job scheduler (with elasticity feature) with 1 trial per experiment (`num_samples=1`). 

## Design Proposal

Below are a few details that we have worked out by reading Ray source code and building prototypes. AdaptDL has three main components: 

1. Scheduler (which includes allocator, optimizer, cluster expander etc.) 

1. The User API 

1. Performance metrics communicator 

All three can be ported using Tune components as described below. 

1. TrialScheduler (Scheduler) 

We are planning to integrate AdaptDL as a custom `TrialScheduler` (e.g., `AdaptDLScheduler`) so that different trials of the same experiment can be scaled to fit the available resources. Users submit `tune.run(..)` or `tune.run_experiments(..)` to run experiments, trials of which are controlled by AdaptDL resource scheduler. Like other trial schedulers (ASHA, PBT), `AdaptDLScheduler` should be compatible with available search algorithms. 
```python
    analysis = tune.run( 
        trainable_cls, 
        scheduler=AdaptDLScheduler(), 
        ..) 
```
We can scale/rescale a trainable by creating new rescaled trainable with the state cloned from the old one. In-memory checkpointing can be used here. We have a working prototype of this technique.

Allocator: The job of the allocator is to come up with a resource (GPUs mostly) allocation plan with the help of the optimizer for all runnable jobs. The scheduler then notices the change in the plan and updates the replicas of the jobs accordingly or starts/stops new jobs. TODO: See if we can use placement groups here to allocate resources in chunks or reclaim old resources. We don’t expect a lot of changes to the allocator and optimizer core logic. 

2. Trainable (User API) 

The trainable for these experiments needs to be distributed-aware, something like what `tune.integration.torch.DistributedTrainableCreator` provides, which takes a train function and number of replicas and creates a Distributed Trainable out of it. We have a slightly modified version of the `TorchTrainable` which we think we can patch dynamically in AdaptDL to avoid creating or extending a brand-new trainable, so users won’t have to learn one more API. The users would write code as they would with the current Callable, but with AdaptDL user API and use `DistributedTrainableCreator` to get a trainable which can be passed to `tune.run`.

> The current `FunctionRunner` used with `DistributedTrainableCreator` lacks restore support. We have added a temporary fix that makes the restore possible.

With 1 and 2 we can scale/rescale multi-replica Tune trials on-the-fly.

3. Metrics Reporting 

We can use/override `tune.report` to piggyback the performance metrics the AdaptDL user library collects. 

4. Cluster Expander

We currently depend heavily on Kubernetes Autoscaler for this functionality.

We'll see if we can use the [existing Ray autoscaler](https://github.com/ray-project/ray/tree/master/python/ray/autoscaler) by integrating it with the scheduler.

 
### Alternatives Considered

N/A

### Performance Implications

Performance impact of AdaptDL optimizer for various Deep Learning workloads/models is available in the linked paper.

### Dependencies

No new dependencies will be added to Tune, except AdaptDL user library that users need to install if they want to use AdaptDL scheduler.

### Engineering Impact

This code would be a pip-installable package maintained by Petuum.

### Platforms and Environments

The proposed scheduler is a Tune-compatible `TrialScheduler` and should support all platforms where Tune's currently supported.

### Best Practices

N/A

### Tutorials and Examples

Here's a complete example of [a distributed trainable](https://github.com/ray-project/ray/blob/master/python/ray/tune/integration/torch.py#L272) modified to use AdaptDL API which can be run with `tune.run`.
```diff
+import adaptdl.torch as adl  

class MyDataset: 
     def __init__(self, xs, ys): 
         self.xs = xs 
         self.ys = ys 

     def __getitem__(self, i): 
         return self.xs[i], self.ys[i] 

     def __len__(self): 
         return len(self.xs) 


def _train_simple(config: Dict, checkpoint_dir: Optional[str] = None): 
     # N is batch size; D_in is input dimension 
     # H is hidden dimension; D_out is output dimension. 
     N, D_in, H, D_out = 8, 5, 5, 5 

     # Create random Tensors to hold inputs and outputs 
     dataset = MyDataset(torch.randn(N, D_in), torch.randn(N, D_out)) 
-    dataloader = DataLoader(dataset) 
+    dataloader = adl.AdaptiveDataLoader(dataset) 

     loss_fn = nn.MSELoss() 

     # Use the nn package to define our model and loss function. 
     model = torch.nn.Sequential( 
         torch.nn.Linear(D_in, H), 
         torch.nn.ReLU(), 
         torch.nn.Linear(H, D_out), 
     ) 

     optimizer = optim.SGD(model.parameters(), lr=0.1) 

     if checkpoint_dir: 
         with open(os.path.join(checkpoint_dir, "checkpoint")) as f: 
             model_state, optimizer_state = torch.load(f) 

         model.load_state_dict(model_state) 
         optimizer.load_state_dict(optimizer_state) 

-    model = DistributedDataParallel(model) 
+    model = adl.AdaptiveDataParallel(model, optimizer) 

-    for epoch in range(config.get("epochs", 10)): 
+    for epoch in adl.remaining_epochs_until(config.get("epochs", 10)): 
         for (x, y) in dataloader: 
             optimizer.zero_grad() 
             output = model(x) 
             loss = loss_fn(output, y) 
             loss.backward() 
             optimizer.step()  

         if epoch % 3 == 0: 
             if config.get("enable_checkpoint", True): 
                 with distributed_checkpoint_dir(step=epoch) as checkpoint_dir: 
                     path = os.path.join(checkpoint_dir, "checkpoint") 
                     torch.save((model.state_dict(), optimizer.state_dict()), 
                                path) 

         tune.report(mean_loss=loss.item()) 

trainable_cls = DistributedTrainableCreator(_train_simple, num_workers=1) 
 
analysis = tune.run( 
     trainable_cls, 
     num_samples=1, 
+    scheduler=adl.AdaptDLScheduler(), 
     stop={"training_iteration": 1}, 
     metric="mean_loss", 
     mode="min") 
```

### Compatibility

The proposal maintains API-level compatibility to Tune, Trainables etc. It does interact with Tune internals components like `TrialRunner`, `RayTrialExecutor` etc. but this is not exposed to the users. We currently expect little to no change to Tune internals, if we manage to use the method patching approach to create the custom AdaptDL trainable. The proposed scheduler should be usable on all platforms where Tune is supported, including the cloud.

### User Impact

The user-facing changes are illustrated in the full example above. The users will have to install AdaptDL user library through `pip`.

## Questions and Discussion Topics

N/A

## Acknowledgement

This template was borrowed from [Ray/RFC](https://github.com/Ray/community)

**Comment by zhisbug** [comment]
- Need to use case about Tune-AdaptDL integration (@aurickq will add later)
- Need to figure autoscaler-Ray integration, maybe we should discuss @richardliaw @aurickq and handle it in another RFC?

**Comment by aurickq** [comment]
There are two primary use cases we are thinking about.

1. Enable a user to launch an individual AdaptDL training job using AWS spot instances in a single command. AdaptDL helps to automatically right-size the cluster for the user according to the trade-off between training goodput and the size (or cost) of the cluster, and dynamically scale the batch size and learning rate. To do this using the current version of AdaptDL [requires significant manual setup](https://adaptdl.readthedocs.io/en/latest/installation/deploy-eks.html) and we are hoping to leverage Ray's infrastructure to make this process seamless and easy to use.
2. Automatically right-size individual trials within a hyper-parameter tuning job, where each trial is an AdaptDL training job, according to (a) each trial's individual goodput scalability characteristics and (b) resource fairness and network interference between multiple trials. AdaptDL may or may not be configured to automatically scale the batch size and learning rate for each trial.

**Comment by valiantljk** [comment]
Interesting idea. Other than Ray tune, any plan to support cluster-wise adaptdl?

**Comment by aurickq** [comment]
> Interesting idea. Other than Ray tune, any plan to support cluster-wise adaptdl?

@valiantljk could you explain a bit what you mean by "cluster-wise adaptdl"?

**Comment by odp** [comment]
The AdaptDL allocator/optimizer comes up with an allocation plan given the running jobs (trials) and available resources. The allocation plan could allocate resources from different nodes in different proportions. We are thinking of using Placement Groups to reserve the allocations (one bundle per node) before they are used by the trial. For example:

```
bundle1 = {"CPU": 2, 'node:192.168.1.X: 0.01}
bundle2 = {"CPU": 1, 'node:192.168.1.Y: 0.01}

pg = placement_group([bundle1, bundle2], strategy="PACK", name="TrialN")

```

Currently the only way you can make a PyTorch distributed trainable use a Placement Group is through the `num_workers_per_host` parameter to `tune.integration.torch.DistributedTrainableCreator` introduced by #11934. It uses a `STRICT_SPREAD` strategy to spread the resources evenly over multiple nodes. 

We need a little finer-grain control over the resource allocation than that. The AdaptDL allocator comes with a specific plan of which replica should go where and can be enforced through a custom Placement Group that we can create from the scheduler or pass it as a factory. Do you think adding an argument `placement_group_factory` or directly a `placement_group` to `DistributedTrainableCreator` makes sense? @richardliaw. Any other suggestions on how to achieve this?

**Comment by zhe-thoughts** [comment]
+ @ericl and @wuisawesome as well. Interesting PG question

**Comment by valiantljk** [comment]
> > Interesting idea. Other than Ray tune, any plan to support cluster-wise adaptdl?
> 
> @valiantljk could you explain a bit what you mean by "cluster-wise adaptdl"?

Ray tune seems more like optimizing a groups of experiments in a single job. 
In case of a multi-tenant ray cluster, or multi-jobs (one ray cluster per job), there might need a global coordinator to facilitate the adaptdl.
